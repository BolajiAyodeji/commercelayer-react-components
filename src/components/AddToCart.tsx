// TODO: add action to onClick event
import React from 'react'

export interface AddToCartProps {
  className?: string
  label?: string
  onClick?: () => void
  disabled?: boolean
}

export default function AddToCart(props) {
  return (
    <button className={props.className}>
      {props.label ? props.label : 'add to cart'}
    </button>
  )
}
