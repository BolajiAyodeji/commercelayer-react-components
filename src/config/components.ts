import PropTypes from 'prop-types'
import childrenTypes from '../utils/childrenTypes'
import {
  TimeFormat,
  BaseInputType,
  GiftCardInputName,
  baseOrderComponentPricePropTypes,
  LineItemType,
  BasePriceType,
  BaseFormatPrice,
  PTLoader,
  BMObject,
  BaseSelectorType,
} from '../@types'
import { ErrorPropTypes } from '../@types/errors'
import { BaseInputComponentPropTypes } from '../@types/index'

const components = {
  AddToCart: {
    displayName: 'AddToCart',
    description: '',
    propTypes: {
      children: PropTypes.func,
      label: PropTypes.string,
      skuCode: PropTypes.string,
      disabled: PropTypes.bool,
    },
    defaultProps: {
      label: 'Add to cart',
    },
  },
  AvailabilityContainer: {
    displayName: 'AvailabilityContainer',
    description: '',
    permittedChildren: ['AvailabilityTemplate', 'ReactNode'],
    propTypes: {
      children: childrenTypes.isRequired,
      skuCode: PropTypes.string,
    },
  },
  AvailabilityTemplate: {
    displayName: 'AvailabilityTemplate',
    description: '',
    propTypes: {
      timeFormat: PropTypes.oneOf<TimeFormat>(['days', 'hours']),
      showShippingMethodName: PropTypes.bool,
      children: PropTypes.func,
    },
    defaultProps: {
      timeFormat: 'days',
      showShippingMethodName: false,
    },
  },
  Checkout: {
    displayName: 'Checkout',
    description: '',
    propTypes: {
      children: PropTypes.func,
      label: PropTypes.string,
    },
    defaultProps: {
      label: 'Checkout',
    },
  },
  CommerceLayer: {
    permittedChildren: [
      'OrderContainer',
      'PriceContainer',
      'GiftCardContainer',
      'ReactNode',
    ],
    propTypes: {
      children: childrenTypes.isRequired,
      accessToken: PropTypes.string.isRequired,
      endpoint: PropTypes.string.isRequired,
      cache: PropTypes.bool,
    },
    defaultProps: {
      cache: false,
    },
  },
  Discount: {
    displayName: 'Discount',
    description: '',
    propTypes: baseOrderComponentPricePropTypes,
    defaultProps: {
      format: 'formatted' as BaseFormatPrice,
    },
  },
  Errors: {
    displayName: 'Errors',
    description: '',
    propTypes: ErrorPropTypes,
    defaultProps: {
      messages: [],
      field: 'base',
    },
  },
  GiftCard: {
    permittedChildren: [
      'GiftCardCurrencySelector',
      'GiftCardInput',
      'Errors',
      'MetadataInput',
      'SubmitButton',
      'ReactNode',
    ],
    displayName: 'GiftCard',
    description: '',
    propTypes: {
      children: childrenTypes.isRequired,
      metadata: PropTypes.objectOf(PropTypes.string),
      onSubmit: PropTypes.func,
    },
    defaultProps: {
      onSubmit: undefined,
      metadata: {},
    },
  },
  GiftCardContainer: {
    permittedChildren: ['GiftCard', 'ReactNode'],
    displayName: 'GiftCardContainer',
    description: '',
    propTypes: {
      children: childrenTypes.isRequired,
    },
  },
  GiftCardCurrencySelector: {
    displayName: 'GiftCardCurrencySelector',
    description: '',
    propTypes: {
      children: PropTypes.func,
      placeholder: PropTypes.exact({
        value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
        label: PropTypes.string,
      }),
      value: PropTypes.string,
      required: PropTypes.bool,
    },
    defaultProps: {
      required: true,
    },
  },
  GiftCardInput: {
    displayName: 'GiftCardInput',
    description: '',
    propTypes: {
      type: PropTypes.oneOf<BaseInputType>([
        'text',
        'email',
        'number',
        'date',
        'checkbox',
      ]).isRequired,
      name: PropTypes.oneOf<GiftCardInputName>([
        'balanceCents',
        'balanceMaxCents',
        'singleUse',
        'rechargeable',
        'imageUrl',
        'expiresAt',
        'referenceOrigin',
        'email',
        'firstName',
        'lastName',
        'reference',
      ]).isRequired,
      children: PropTypes.func,
      placeholder: PropTypes.string,
    },
  },
  GiftCardPrice: {
    displayName: 'GiftCardPrice',
    description: '',
    propTypes: baseOrderComponentPricePropTypes,
  },
  ItemContainer: {
    permittedChildren: [
      'PriceContainer',
      'VariantContainer',
      'SkuOptionContainer',
      'QuantitySelector',
      'AddToCart',
      'AvailabilityContainer',
      'ReactNode',
    ],
    displayName: 'ItemContainer',
    description: '',
    propTypes: {
      children: childrenTypes.isRequired,
      skuCode: PropTypes.string,
    },
  },
  LineItem: {
    permittedChildren: [
      'LineItemImage',
      'LineItemName',
      'LineItemOptions',
      'LineItemQuantity',
      'LineItemPrice',
      'LineItemRemove',
      'Errors',
      'ReactNode',
    ],
    displayName: 'LineItem',
    description: '',
    propTypes: {
      children: childrenTypes.isRequired,
      type: PropTypes.oneOf<LineItemType>([
        'skus',
        'gift_cards',
        'shipments',
        'paymentMethods',
        'promotions',
      ]),
    },
    defaultProps: {
      type: 'skus',
    },
  },
  LineItemImage: {
    displayName: 'LineItemImage',
    description: '',
    propTypes: {
      width: PropTypes.number,
      src: PropTypes.string,
      children: PropTypes.func,
    },
  },
  LineItemName: {
    displayName: 'LineItemName',
    description: '',
    propTypes: {
      children: PropTypes.func,
      label: PropTypes.string,
    },
  },
  LineItemOption: {
    displayName: 'LineItemOption',
    description: '',
    propTypes: {
      name: PropTypes.string.isRequired,
      children: PropTypes.func,
      valueClassName: PropTypes.string,
      keyClassName: PropTypes.string,
      keyId: PropTypes.string,
      keyStyle: PropTypes.object,
    },
  },
  LineItemOptions: {
    permittedChildren: ['LineItemOption', 'ReactNode'],
    displayName: 'LineItemOptions',
    description: '',
    propTypes: {
      name: PropTypes.string.isRequired,
      children: childrenTypes.isRequired,
      showName: PropTypes.bool,
    },
    defaultProps: {
      showName: true,
    },
  },
  LineItemPrice: {
    displayName: 'LineItemPrice',
    description: '',
    propTypes: {
      ...baseOrderComponentPricePropTypes,
      type: PropTypes.oneOf<BasePriceType>(['total', 'unit', 'option']),
    },
    defaultProps: {
      format: 'formatted' as BaseFormatPrice,
      type: 'total' as BasePriceType,
    },
  },
  LineItemQuantity: {
    displayName: 'LineItemQuantity',
    description: '',
    propTypes: {
      children: PropTypes.func,
      max: PropTypes.number,
      disabled: PropTypes.bool,
    },
    defaultProps: {
      max: 50,
    },
  },
  LineItemRemove: {
    displayName: 'LineItemRemove',
    description: '',
    propTypes: {
      children: PropTypes.func,
      label: PropTypes.string,
    },
    defaultProps: {
      label: 'remove',
    },
  },
  LineItemsContainer: {
    permittedChildren: ['LineItemsCount', 'LineItem', 'ReactNode'],
    displayName: 'LineItemsContainer',
    description: '',
    propTypes: {
      children: childrenTypes.isRequired,
      filters: PropTypes.object,
      loader: PTLoader,
    },
    defaultProps: {
      filters: {},
      loader: 'Loading...',
    },
  },
  LineItemsCount: {
    displayName: 'LineItemsCount',
    description: '',
    propTypes: {
      children: PropTypes.func,
      id: PropTypes.string,
      className: PropTypes.string,
      name: PropTypes.string,
      style: PropTypes.object,
    },
  },
  MetadataInput: {
    displayName: 'MetadataInput',
    description: '',
    propTypes: BaseInputComponentPropTypes,
  },
  OrderContainer: {
    permittedChildren: [
      'ItemContainer',
      'LineItemsContainer',
      'SubTotal',
      'Discount',
      'Shipping',
      'Taxes',
      'GiftCardPrice',
      'Total',
      'Checkout',
      'GiftCardContainer',
      'ReactNode',
    ],
    displayName: 'OrderContainer',
    description: '',
    propTypes: {
      children: childrenTypes.isRequired,
      persistKey: PropTypes.string.isRequired,
      metadata: BMObject,
    },
    defaultProps: {
      metadata: {},
    },
  },
  Price: {
    displayName: 'Price',
    description: '',
    propTypes: {
      children: PropTypes.func,
      compareClassName: PropTypes.string,
      skuCode: PropTypes.string,
      showCompare: PropTypes.bool,
    },
    defaultProps: {
      skuCode: '',
    },
  },
  PriceContainer: {
    permittedChildren: ['Price', 'ReactNode'],
    displayName: 'PriceContainer',
    description: '',
    propTypes: {
      children: childrenTypes.isRequired,
      skuCode: PropTypes.string,
      loader: PTLoader,
      perPage: PropTypes.number,
      filters: PropTypes.object,
    },
    defaultProps: {
      perPage: 10,
      filters: {},
      loader: 'Loading...',
      skuCode: '',
    },
  },
  PriceTemplate: {
    displayName: 'PriceTemplate',
    description: '',
    propTypes: {
      formattedAmount: PropTypes.string,
      formattedCompare: PropTypes.string,
    },
    defaultProps: {
      formattedAmount: '',
      formattedCompare: '',
    },
  },
  QuantitySelector: {
    displayName: 'QuantitySelector',
    description: '',
    propTypes: {
      children: PropTypes.func,
      min: PropTypes.string,
      max: PropTypes.string,
      value: PropTypes.string,
      skuCode: PropTypes.string,
      disabled: PropTypes.bool,
    },
    defaultProps: {
      min: '1',
    },
  },
  Shipping: {
    displayName: 'Shipping',
    description: '',
    propTypes: baseOrderComponentPricePropTypes,
  },
  SkuOption: {
    permittedChildren: ['SkuOptionInput', 'ReactNode'],
    displayName: 'SkuOption',
    description: '',
    propTypes: {
      children: childrenTypes.isRequired,
      name: PropTypes.string.isRequired,
    },
  },
  SkuOptionInput: {
    displayName: 'SkuOptionInput',
    description: '',
    propTypes: BaseInputComponentPropTypes,
  },
  SkuOptionsContainer: {
    permittedChildren: ['SkuOption', 'ReactNode'],
    displayName: 'SkuOptionsContainer',
    description: '',
    propTypes: {
      children: childrenTypes.isRequired,
      skuCode: PropTypes.string,
    },
  },
  SubmitButton: {
    displayName: 'SubmitButton',
    description: '',
    propTypes: {
      children: PropTypes.func,
      label: PropTypes.string,
    },
    defaultProps: {
      label: 'Submit',
    },
  },
  SubTotal: {
    displayName: 'SubTotal',
    description: '',
    propTypes: baseOrderComponentPricePropTypes,
    defaultProps: {
      format: 'formatted' as BaseFormatPrice,
    },
  },
  Taxes: {
    displayName: 'Taxes',
    description: '',
    propTypes: baseOrderComponentPricePropTypes,
    defaultProps: {
      format: 'formatted' as BaseFormatPrice,
    },
  },
  Total: {
    displayName: 'Total',
    description: '',
    propTypes: baseOrderComponentPricePropTypes,
    defaultProps: {
      format: 'formatted' as BaseFormatPrice,
    },
  },
  VariantContainer: {
    permittedChildren: ['VariantSelector', 'ReactNode'],
    displayName: 'VariantContainer',
    description: '',
    propTypes: {
      children: childrenTypes.isRequired,
      skuCode: PropTypes.string,
      filters: PropTypes.object,
    },
    defaultProps: {
      skuCode: '',
      filters: {},
    },
  },
  VariantSelector: {
    displayName: 'VariantSelector',
    description: '',
    propTypes: {
      skuCodes: PropTypes.arrayOf(
        PropTypes.exact({
          label: PropTypes.string.isRequired,
          code: PropTypes.string.isRequired,
        }).isRequired
      ).isRequired,
      name: PropTypes.string,
      children: PropTypes.func,
      type: PropTypes.oneOf<BaseSelectorType>(['select', 'radio']),
      loader: PropTypes.element,
      placeholder: PropTypes.string,
      skuCode: PropTypes.string,
    },
    defaultProps: {
      placeholder: 'select an variant',
      type: 'select' as BaseSelectorType,
    },
  },
  VariantTemplate: {
    displayName: 'VariantTemplate',
    description: '',
    propTypes: {
      variants: PropTypes.object.isRequired,
      onChange: PropTypes.func,
      skuCodes: PropTypes.arrayOf(
        PropTypes.exact({
          label: PropTypes.string.isRequired,
          code: PropTypes.string.isRequired,
        }).isRequired
      ).isRequired,
      name: PropTypes.string,
      children: PropTypes.func,
      type: PropTypes.oneOf<BaseSelectorType>(['select', 'radio']),
      loader: PropTypes.element,
      placeholder: PropTypes.string,
      skuCode: PropTypes.string,
    },
  },
}

export default components
