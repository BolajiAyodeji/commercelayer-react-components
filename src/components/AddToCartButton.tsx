import React, {
  FunctionComponent,
  useContext,
  ReactNode,
  PropsWithoutRef,
} from 'react'
import Parent from './utils/Parent'
import OrderContext from '../context/OrderContext'
import _ from 'lodash'
import ItemContext from '../context/ItemContext'
import getCurrentItemKey from '../utils/getCurrentItemKey'
import components from '../config/components'

const propTypes = components.AddToCartButton.propTypes
const defaultProps = components.AddToCartButton.defaultProps
const displayName = components.AddToCartButton.displayName

type ChildrenFunctionProps = {
  handleClick: () => void
  label?: string
  skuCode?: string
  disabled?: boolean
} & PropsWithoutRef<JSX.IntrinsicElements['button']>

type AddToCartButtonProps = Omit<ChildrenFunctionProps, 'handleClick'> & {
  children?: (props: ChildrenFunctionProps) => ReactNode
} & PropsWithoutRef<JSX.IntrinsicElements['button']>

const AddToCartButton: FunctionComponent<AddToCartButtonProps> = (props) => {
  const { label = 'Add to cart', children, skuCode, disabled, ...p } = props
  const { addToCart } = useContext(OrderContext)
  const {
    item,
    items,
    quantity,
    option,
    prices,
    lineItems,
    lineItem,
    skuCode: itemSkuCode,
  } = useContext(ItemContext)
  const sCode =
    !_.isEmpty(items) && skuCode
      ? items[skuCode]?.code
      : skuCode || getCurrentItemKey(item) || (itemSkuCode as string)
  const handleClick = (): void => {
    const qty = quantity[sCode]
    const opt = option[sCode]
    const customLineItem = !_.isEmpty(lineItem) ? lineItem : lineItems[sCode]
    if (addToCart)
      addToCart({
        skuCode: sCode,
        skuId: item[sCode]?.id,
        quantity: qty,
        option: opt,
        lineItem: customLineItem,
      })
  }
  const autoDisabled = disabled || !prices[sCode] || !sCode
  const parentProps = {
    handleClick,
    disabled: disabled || autoDisabled,
    label,
    ...props,
  }
  return children ? (
    <Parent {...parentProps}>{children}</Parent>
  ) : (
    <button disabled={autoDisabled} onClick={handleClick} {...p}>
      {label}
    </button>
  )
}

AddToCartButton.propTypes = propTypes
AddToCartButton.defaultProps = defaultProps
AddToCartButton.displayName = displayName

export default AddToCartButton
