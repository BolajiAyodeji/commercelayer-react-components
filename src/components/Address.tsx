import React, {
  FunctionComponent,
  Fragment,
  useContext,
  ReactNode,
  useState,
  useEffect,
} from 'react'
import AddressChildrenContext from '#context/AddressChildrenContext'
import components from '#config/components'
import CustomerContext from '#context/CustomerContext'
import BillingAddressContext from '#context/BillingAddressContext'
import ShippingAddressContext from '#context/ShippingAddressContext'
import { AddressCollection } from '@commercelayer/js-sdk'
import _ from 'lodash'
import AddressContext from '#context/AddressContext'

const propTypes = components.Address.propTypes

type Props = {
  children: ReactNode
  selectedClassName?: string
  onSelect?: () => void
  addresses?: AddressCollection[]
} & JSX.IntrinsicElements['div']

const Address: FunctionComponent<Props> = (props) => {
  const {
    children,
    className,
    selectedClassName,
    onSelect,
    addresses = [],
    ...p
  } = props
  const { addresses: addressesContext } = useContext(CustomerContext)
  const { setBillingAddress, billingCustomerAddressId } = useContext(
    BillingAddressContext
  )
  const { setShippingAddress, shippingCustomerAddressId } = useContext(
    ShippingAddressContext
  )
  const {
    billingAddress,
    shippingAddress,
    shipToDifferentAddress,
  } = useContext(AddressContext)
  const [selected, setSelected] = useState<null | number>(null)
  const items = !_.isEmpty(addresses)
    ? addresses
    : (addressesContext && addressesContext) || []
  useEffect(() => {
    if (items) {
      items.map((address, k) => {
        if (
          billingCustomerAddressId ||
          shippingCustomerAddressId === undefined
        ) {
          const preselected =
            address.customerAddressId === billingCustomerAddressId
          if (preselected && selected === null && _.isEmpty(billingAddress))
            setSelected(k)
          else if (
            (preselected || selected !== null) &&
            !_.isEmpty(billingAddress)
          )
            setSelected(null)
        }
        if (
          (shipToDifferentAddress && shippingCustomerAddressId) ||
          billingCustomerAddressId === undefined
        ) {
          const preselected =
            address.customerAddressId === shippingCustomerAddressId
          if (preselected && selected === null && _.isEmpty(shippingAddress))
            setSelected(k)
          else if (
            (preselected || selected !== null) &&
            !_.isEmpty(shippingAddress)
          )
            setSelected(null)
        }
      })
    }
  }, [billingAddress, shippingAddress, addresses])
  const handleSelect = async (
    k: number,
    addressId: string,
    customerAddressId: string
  ) => {
    if (k !== selected) {
      setSelected(k)
      setBillingAddress &&
        (await setBillingAddress(addressId, { customerAddressId }))
      setShippingAddress &&
        (await setShippingAddress(addressId, { customerAddressId }))
      onSelect && onSelect()
    }
  }
  const components = items.map((address, k) => {
    const addressProps = {
      address,
    }
    const addressSelectedClass =
      selected === k ? `${className} ${selectedClassName}` : className
    const customerAddressId: string = address?.customerAddressId || ''
    return (
      <AddressChildrenContext.Provider key={k} value={addressProps}>
        <div
          className={addressSelectedClass}
          onClick={() => handleSelect(k, address.id, customerAddressId)}
          {...p}
        >
          {children}
        </div>
      </AddressChildrenContext.Provider>
    )
  })
  return <Fragment>{components}</Fragment>
}

Address.propTypes = propTypes

export default Address
