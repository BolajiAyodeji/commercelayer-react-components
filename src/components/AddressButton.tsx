import React, { FunctionComponent, useContext } from 'react'
import Parent from './utils/Parent'
import components from '../config/components'
import { FunctionChildren } from '../typings/index'
import AddressContext from '../context/AddressContext'
import _ from 'lodash'
import { fieldsExist } from '../utils/validateFormFields'
import { addressFields } from '../reducers/AddressReducer'

const propTypes = components.AddressButton.propTypes
const defaultProps = components.AddressButton.defaultProps
const displayName = components.AddressButton.displayName

type AddressButtonChildrenProps = FunctionChildren<
  Omit<AddressButtonProps, 'children'>
>

type AddressButtonProps = {
  children?: AddressButtonChildrenProps
  label?: string
  onClick?: () => void
} & JSX.IntrinsicElements['button']

const AddressButton: FunctionComponent<AddressButtonProps> = (props) => {
  const {
    children,
    label = 'Continue to delivery',
    resource,
    disabled,
    onClick,
    ...p
  } = props
  const {
    errors,
    billingAddress,
    shipToDifferentAddress,
    shippingAddress,
    saveAddresses,
  } = useContext(AddressContext)
  let disable =
    disabled ||
    !_.isEmpty(errors) ||
    _.isEmpty(billingAddress) ||
    (shipToDifferentAddress && _.isEmpty(shippingAddress))
  if (_.isEmpty(errors) && !_.isEmpty(billingAddress)) {
    disable = billingAddress && fieldsExist(billingAddress)
  }
  if (
    _.isEmpty(errors) &&
    shipToDifferentAddress
    // !_.isEmpty(shippingAddress)
  ) {
    disable =
      shippingAddress &&
      fieldsExist(shippingAddress, _.without(addressFields, 'customer_email'))
  }
  const handleClick = () => {
    if (_.isEmpty(errors) && !disable) {
      saveAddresses()
      onClick && onClick()
    }
  }
  const parentProps = {
    ...p,
    label,
    resource,
    handleClick,
    disabled: disable,
  }
  return children ? (
    <Parent {...parentProps}>{children}</Parent>
  ) : (
    <button type="button" disabled={disable} onClick={handleClick} {...p}>
      {label}
    </button>
  )
}

AddressButton.propTypes = propTypes
AddressButton.defaultProps = defaultProps
AddressButton.displayName = displayName

export default AddressButton
