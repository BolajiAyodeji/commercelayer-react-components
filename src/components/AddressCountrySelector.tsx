import React, { FunctionComponent, useContext, useEffect } from 'react'
import BaseSelect from './utils/BaseSelect'
import countriesOptions from '../config/countries.json'
import components from '../config/components'
import { AddressCountrySelectName, BaseSelectComponentProps } from '../typings'
import BillingAddressContext from '../context/BillingAddressContext'
import ShippingAddressContext from '../context/ShippingAddressContext'

const propTypes = components.AddressCountrySelector.propTypes
const defaultProps = components.AddressCountrySelector.defaultProps
const displayName = components.AddressCountrySelector.displayName

type AddressCountrySelectorProps = Omit<
  BaseSelectComponentProps,
  'options' | 'name'
> & {
  name: AddressCountrySelectName
  required?: boolean
} & Pick<JSX.IntrinsicElements['select'], 'className' | 'id' | 'style'>

const AddressCountrySelector: FunctionComponent<AddressCountrySelectorProps> = (
  props
) => {
  const { required = true, value, name } = props
  const billingAddress = useContext(BillingAddressContext)
  const shippingAddress = useContext(ShippingAddressContext)
  useEffect(() => {
    if (value && billingAddress?.setValue) {
      billingAddress.setValue(name, value)
    }
    if (value && shippingAddress?.setValue) {
      shippingAddress.setValue(name, value)
    }
  }, [value])
  return (
    <BaseSelect
      ref={billingAddress?.validation || shippingAddress?.validation}
      required={required}
      options={countriesOptions}
      {...props}
    />
  )
}

AddressCountrySelector.propTypes = propTypes
AddressCountrySelector.defaultProps = defaultProps
AddressCountrySelector.displayName = displayName

export default AddressCountrySelector
