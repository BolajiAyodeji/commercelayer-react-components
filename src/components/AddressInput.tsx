import React, { FunctionComponent, useContext, useEffect } from 'react'
import BaseInput from './utils/BaseInput'
import components from '../config/components'
import { BaseInputComponentProps, AddressInputName } from '../typings'
import BillingAddressContext from '../context/BillingAddressContext'
import ShippingAddressContext from '../context/ShippingAddressContext'

const propTypes = components.AddressInput.propTypes
const displayName = components.AddressInput.displayName

export type AddressInputProps = {
  name: AddressInputName
} & Omit<BaseInputComponentProps, 'name'> &
  JSX.IntrinsicElements['input'] &
  JSX.IntrinsicElements['textarea']

const AddressInput: FunctionComponent<AddressInputProps> = (props) => {
  const { placeholder = '', required = true, value, ...p } = props
  const billingAddress = useContext(BillingAddressContext)
  const shippingAddress = useContext(ShippingAddressContext)
  useEffect(() => {
    if (value && billingAddress?.setValue) {
      billingAddress.setValue(p.name, value)
    }
    if (value && shippingAddress?.setValue) {
      shippingAddress.setValue(p.name, value)
    }
  }, [value])
  return (
    <BaseInput
      ref={billingAddress?.validation || shippingAddress?.validation}
      required={required}
      placeholder={placeholder}
      defaultValue={value}
      {...p}
    />
  )
}

AddressInput.propTypes = propTypes
AddressInput.displayName = displayName

export default AddressInput
