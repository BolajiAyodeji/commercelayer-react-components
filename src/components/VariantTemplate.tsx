import React, { Fragment, FunctionComponent } from 'react'
import _ from 'lodash'
import { PropsType } from '../utils/PropsType'
import components from '../config/components'

const propTypes = components.VariantTemplate.propTypes
const displayName = components.VariantTemplate.displayName

export type VariantTemplateProps = PropsType<typeof propTypes> &
  JSX.IntrinsicElements['input'] &
  JSX.IntrinsicElements['select']

const VariantTemplate: FunctionComponent<VariantTemplateProps> = (props) => {
  const {
    id,
    variants,
    type,
    placeholder,
    skuCodes,
    skuCode,
    onChange,
    ...prs
  } = props
  const vars = _.keys(variants).map((v, k) => {
    const checked = skuCode === v
    return type === 'select' ? (
      <option
        key={variants[v].id}
        data-sku-id={variants[v].id}
        value={variants[v].code}
      >
        {skuCodes.length > 0 ? skuCodes[k].label : variants[v].name}
      </option>
    ) : (
      <Fragment key={variants[v].id}>
        <input
          id={id && `${id}-${k}`}
          defaultChecked={checked}
          type="radio"
          value={variants[v].code}
          onChange={(e): void =>
            onChange && onChange(e.target.value, variants[v].id)
          }
          {...prs}
        />
        {skuCodes.length > 0 ? skuCodes[k].label : variants[v].name}
      </Fragment>
    )
  })
  if (type === 'select') {
    return (
      <select
        id={id}
        onChange={(e): void => {
          const v = e.target.value
          const i = e.target.selectedIndex
          const id = e.target[i].dataset.skuId
          onChange && onChange(v, id)
        }}
        value={skuCode || ''}
        {...prs}
      >
        <option disabled={!!skuCode}>{placeholder}</option>
        {vars}
      </select>
    )
  }
  return <Fragment>{vars}</Fragment>
}

VariantTemplate.propTypes = propTypes
VariantTemplate.displayName = displayName

export default VariantTemplate
