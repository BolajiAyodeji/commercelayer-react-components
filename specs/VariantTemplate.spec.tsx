import React from 'react'
import VariantTemplate from '../src/components/VariantTemplate'
import renderer from 'react-test-renderer'
import components from '../src/config/components'

const propTypes = components.VariantTemplate.propTypes

test('<VariantTemplate/>', () => {
  expect.assertions(9)
  const component = renderer.create(
    <VariantTemplate
      skuCodes={[
        {
          label: '6 months',
          code: 'BABYONBU000000E63E746MXX',
        },
      ]}
      variants={{}}
    />
  )
  const tree = component.toJSON()
  const root = component.toTree()
  const proptypes = root.type['propTypes']
  expect(tree).toMatchSnapshot()
  expect(proptypes.children).toBe(propTypes.children)
  expect(proptypes.placeholder).toBe(propTypes.placeholder)
  expect(proptypes.skuCodes).toBe(propTypes.skuCodes)
  expect(proptypes.name).toBe(propTypes.name)
  expect(proptypes.type).toBe(propTypes.type)
  expect(proptypes.skuCode).toBe(propTypes.skuCode)
  expect(proptypes.variants).toBe(propTypes.variants)
  expect(proptypes.onChange).toBe(propTypes.onChange)
})
