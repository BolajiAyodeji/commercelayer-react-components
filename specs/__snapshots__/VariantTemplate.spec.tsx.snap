// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VariantTemplate check type radio /> 1`] = `null`;

exports[`<VariantTemplate check type select /> 1`] = `null`;

exports[`<VariantTemplate/> 1`] = `null`;
